src/main/java/com/mongodb/kafka/connect/source/MongoCopyDataManager.java

```diff
@@ -166,20 +185,32 @@ private void putToQueue(final BsonDocument bsonDocument) {
     cfg.getPipeline(COPY_EXISTING_PIPELINE_CONFIG).map(pipeline::addAll);
     pipeline.add(
         BsonDocument.parse(
-            "{$replaceRoot: {newRoot: {"
-                + "_id: {_id: '$_id', copyingData: true}, "
-                + "operationType: 'insert', "
-                + "ns: {db: '"
-                + namespace.getDatabaseName()
-                + "', coll: '"
-                + namespace.getCollectionName()
-                + "'}, "
-                + "documentKey: {_id: '$_id'}, "
-                + "fullDocument: '$$ROOT'}}}"));
+            format(PIPELINE_TEMPLATE, namespace.getDatabaseName(), namespace.getCollectionName())));
     cfg.getPipeline().map(pipeline::addAll);
     return pipeline;
   }
 
+  static RawBsonDocument convertDocument(final RawBsonDocument original) {
+    ByteBuffer sourceBuffer = original.getByteBuffer().asNIO();
+    BsonBinaryReader reader = new BsonBinaryReader(sourceBuffer);
+    int currentPosition = 0;
+    reader.readStartDocument();
+    while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+      if (reader.readName().equals(NAMESPACE_FIELD)) {
+        currentPosition++; // Adjust the current position to include the bson type
+        byte[] sourceBytes = sourceBuffer.array();
+        // Convert the namespace field in situ
+        for (byte namespaceByte : NAMESPACE_BYTES) {
```

Comment:
I probably completely misunderstand what is going on here, so I apologize for this question if it makes no sense: why cannot we just use `"ns"` instead of `"__"` in `PIPELINE_TEMPLATE`?

Reply:
See the related issues in KAFKA-201 for the background information.

Reply:
Thank you for explaining this to me!

Reply:
Hi @rozza 
Could you please give more details?
Maybe make sense to add a comment in the code.

I'm asking in old MR because here was broken backward compatibility in the case of `ns` fields using in `pipeline` config.

P.s: I saw the related issues in KAFKA-201, but it doesn't have much description :(

Reply:
Here is example which could be added to `MongoSourceConnectorIntegrationTest`


```java
  @Test
  @DisplayName("Ensure source loads and stream data from certain collections")
  void testSourceLoadsAndStreamDataFromCertainCollections() {
    assumeTrue(isGreaterThanThreeDotSix());
    MongoDatabase db = getDatabaseWithPostfix();
    MongoCollection<Document> coll1 = db.getCollection("coll1");
    MongoCollection<Document> coll2 = db.getCollection("coll2");
    MongoCollection<Document> coll3 = db.getCollection("coll3");

    insertMany(rangeClosed(1, 50), coll1);
    insertMany(rangeClosed(1, 50), coll2);
    insertMany(rangeClosed(1, 50), coll3);

    Properties sourceProperties = new Properties();
    sourceProperties.put(MongoSourceConfig.COPY_EXISTING_CONFIG, "true");
    String namespaceRegex = String.format("(%s\\.coll(1|2))", db.getName());
    sourceProperties.put(MongoSourceConfig.COPY_EXISTING_NAMESPACE_REGEX_CONFIG, namespaceRegex);
    sourceProperties.put(PIPELINE_CONFIG, "[{\"$match\": {\"ns.coll\": {\"$regex\": /^coll(1|2)$/}}}]");
    /* 
    Works with such pipeline
    sourceProperties.put(
        PIPELINE_CONFIG,
        "[{\"$match\": {\"$or\": [{\"ns.coll\": {\"$regex\": /^coll(1|2)$/}}, {\"__.coll\": {\"$regex\": /^coll(1|2)$/}}]}}]");
     */

    addSourceConnector(sourceProperties);

    insertMany(rangeClosed(51, 100), coll1);
    insertMany(rangeClosed(51, 100), coll2);
    insertMany(rangeClosed(51, 100), coll3);

    assertProduced(createInserts(1, 100), coll1);
    assertProduced(createInserts(1, 100), coll2);
    // TODO: assertProducedNone(coll3)
  }
```

Reply:
I've created a separate ticket [KAFKA-217](https://jira.mongodb.org/projects/KAFKA/issues/KAFKA-217)

Reply:
Thanks @ortex that is indeed an unintended side effect.